### PR TITLE
Set up base subgraph project

### PR DESCRIPTION
**Pull request by mtullyoc**
Resolves:

[Linear Issue BAL-33 ](https://linear.app/bleu-llc/issue/BAL-33/set-up-base-subgraph-with-abi-file-and-sc-address)
- Set up a base subgraph project

[Linear Issue BAL-34](https://linear.app/bleu-llc/issue/BAL-34/create-base-entities-and-define-graphql-schema)
- Created schema.graphql
- Defined Pool and PoolMetadataUpdated entities

[Linear Issue BAL-35](https://linear.app/bleu-llc/issue/BAL-35/create-subgraph-manifest-and-define-handlers)
- Created networks.yaml to define the different addresses across the networks
- Created generate-manifests.ts to scaffold the subgraph.[network].yaml

Closes BAL-33, BAL-34 and BAL-35

### CODE REVIEW

File: packages/subgraph/generated/PoolMetadataRegistry/PoolMetadataRegistry.ts [hunk 1]
```diff
@@ -0,0 +1,145 @@
+// THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (mendesfabio) [inline packages/subgraph/generated/PoolMetadataRegistry/PoolMetadataRegistry.ts:1]: can you add `generated` folder to `.gitignore`?

File: packages/subgraph/schema.graphql [hunk 2]
```diff
@@ -0,0 +1,8 @@
+type PoolMetadataUpdated @entity(immutable: true) {
```
Comment (mendesfabio) [inline packages/subgraph/schema.graphql:1]: I believe we need another entity `Pool` that registers the latest `metadataCID` - that's the one apps will query.

File: packages/subgraph/schema.graphql [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+  id: Bytes!
+  poolId: Bytes! # bytes32
```
Comment (mendesfabio) [inline packages/subgraph/schema.graphql:3]: can you add proper documentation for every entity and field? example: https://github.com/balancer/gauges-subgraph/blob/master/schema.graphql

File: packages/subgraph/src/pool-metadata-registry.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { PoolMetadataUpdated as PoolMetadataUpdatedEvent } from "../generated/PoolMetadataRegistry/PoolMetadataRegistry"
+import { PoolMetadataUpdated } from "../generated/schema"
+
+export function handlePoolMetadataUpdated(
```
Comment (mendesfabio) [inline packages/subgraph/src/pool-metadata-registry.ts:4]: once you create the new entity, remember to update this handler.

File: packages/subgraph/subgraph.yaml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+specVersion: 0.0.5
+schema:
+  file: ./schema.graphql
+dataSources:
+  - kind: ethereum
+    name: PoolMetadataRegistry
+    network: goerli
+    source:
+      address: "0xEBfADf723E077c80f6058dC9c9202bB613DE07cf"
```
Comment (mendesfabio) [inline packages/subgraph/subgraph.yaml:9]: not sure if we will be able to test this contract - get the latest one with @devjoaov

File: packages/subgraph/tests/pool-metadata-registry.test.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+  afterAll
+} from "matchstick-as/assembly/index"
+import { Bytes } from "@graphprotocol/graph-ts"
+import { PoolMetadataUpdated } from "../generated/schema"
+import { PoolMetadataUpdated as PoolMetadataUpdatedEvent } from "../generated/PoolMetadataRegistry/PoolMetadataRegistry"
+import { handlePoolMetadataUpdated } from "../src/pool-metadata-registry"
+import { createPoolMetadataUpdatedEvent } from "./pool-metadata-registry-utils"
+
+// Tests structure (matchstick-as >=0.5.0)
+// https://thegraph.com/docs/en/developer/matchstick/#tests-structure-0-5-0
```
Comment (mendesfabio) [inline packages/subgraph/tests/pool-metadata-registry.test.ts:16]: can you remove the comments?

File: packages/subgraph/package.json [hunk 7]
```diff
@@ -0,0 +1,18 @@
+{
+  "name": "pool-matadata-registry-graph",
```
Comment (mendesfabio) [inline packages/subgraph/package.json:2]:
```suggestion
  "name": "pool-metadata-registry-subgraph",
```

File: packages/subgraph/package.json [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  "license": "UNLICENSED",
+  "scripts": {
+    "codegen": "graph codegen",
+    "build": "graph build",
+    "deploy": "graph deploy --node https://api.studio.thegraph.com/deploy/ pool-matadata-registry-graph",
+    "create-local": "graph create --node http://localhost:8020/ pool-matadata-registry-graph",
+    "remove-local": "graph remove --node http://localhost:8020/ pool-matadata-registry-graph",
+    "deploy-local": "graph deploy --node http://localhost:8020/ --ipfs http://localhost:5001 pool-matadata-registry-graph",
```
Comment (mendesfabio) [inline packages/subgraph/package.json:10]: were you able to deploy locally? can we have a README with `anvil`, `graph-node`, etc, set up?

File: packages/subgraph/networks.json [hunk 9]
```diff
@@ -0,0 +1,8 @@
+{
+  "goerli": {
+    "PoolMetadataRegistry": {
+      "address": "0xEBfADf723E077c80f6058dC9c9202bB613DE07cf",
+      "startBlock": 8583917
+    }
+  }
+}
```
Comment (mendesfabio) [inline packages/subgraph/networks.json:8]: is this intended to generate manifests automatically? what's the script that does this?

File: packages/subgraph/tests/pool-metadata-registry.test.ts [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+describe("Describe entity assertions", () => {
+  beforeAll(() => {
+    let poolId = Bytes.fromI32(1234567890)
+    let metadataCID = "Example string value"
```
Comment (mendesfabio) [inline packages/subgraph/tests/pool-metadata-registry.test.ts:21]: this should look like a IPFS CID e.g. `QmPq9...`

File: packages/subgraph/tests/pool-metadata-registry.test.ts [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+import { PoolMetadataUpdated as PoolMetadataUpdatedEvent } from "../generated/PoolMetadataRegistry/PoolMetadataRegistry"
+import { handlePoolMetadataUpdated } from "../src/pool-metadata-registry"
+import { createPoolMetadataUpdatedEvent } from "./pool-metadata-registry-utils"
+
+// Tests structure (matchstick-as >=0.5.0)
+// https://thegraph.com/docs/en/developer/matchstick/#tests-structure-0-5-0
+
+describe("Describe entity assertions", () => {
+  beforeAll(() => {
+    let poolId = Bytes.fromI32(1234567890)
```
Comment (mendesfabio) [inline packages/subgraph/tests/pool-metadata-registry.test.ts:20]: this should look like a real poolId

**Review by mendesfabio — CHANGES_REQUESTED** [review]
looking good - left some suggestions.

File: packages/subgraph/subgraph.yaml (hunk 5 above)
Comment (ribeirojose) [inline packages/subgraph/subgraph.yaml:9]: I'm not knowledgeable in subgraph, but I feel this file should be automatically generated, no?

**Review by ribeirojose — DISMISSED** [review]
😪  so much boilerplate

File: packages/subgraph/package.json (hunk 7 above)
Comment (mtullyoc) [reply packages/subgraph/package.json:2]: Should we do it as in contracts/package.json?

```
  "name": "@balancer-pool-metadata/subgraph",
  "version": "1.0.0",
```

File: packages/subgraph/schema.graphql (hunk 2 above)
Comment (mtullyoc) [reply packages/subgraph/schema.graphql:1]: WDYT we add 

`metadataUpdates: [PoolMetadataUpdated!] @derivedFrom(field: "poolId") `

to Pool entity?

File: packages/subgraph/subgraph.yaml (hunk 5 above)
Comment (mtullyoc) [reply packages/subgraph/subgraph.yaml]: On Balancer's repo there's a script generate-manifests to produce subgraph.yaml from a template. I made the same here and deleted this. 😉

File: packages/subgraph/tsconfig.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+  "compilerOptions": {
+    "lib": ["DOM","ES2018"],
+    "types": ["@graphprotocol/graph-ts", "node"],
+    "skipLibCheck": true,                     /* Skip type checking of declaration files. */
+    "forceConsistentCasingInFileNames": true,  /* Disallow inconsistently-cased references to the same file. */
+    "strictNullChecks": true
+  },
+  "exclude": [
+    "src/types"
+  ]
```
Comment (ribeirojose) [inline packages/subgraph/tsconfig.json:11]: ❓

Comment (ribeirojose) [inline packages/subgraph/tsconfig.json:11]: I think the last version inherited from the subgraph tsconfig, any reason to change that?

File: packages/subgraph/scripts/generate-manifests.ts [hunk 13]
```diff
@@ -0,0 +1,25 @@
+import yaml = require('js-yaml');
+
+import Handlebars = require('handlebars');
```
Comment (ribeirojose) [inline packages/subgraph/scripts/generate-manifests.ts:3]: THAT'S the lib name I was trying to remember a couple days ago 😆 @mendesfabio

File: packages/subgraph/networks.yaml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+# mainnet:
+#   network: mainnet
+#   EventEmitter:
+#     address: "0x000000000000000000000000000000000000000"
+#     startBlock: 16419620
+goerli:
```
Comment (ribeirojose) [inline packages/subgraph/networks.yaml:6]: 👍

File: packages/subgraph/manifest.template.yaml [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+      language: wasm/assemblyscript
+      entities:
+        - PoolMetadataUpdated
+      abis:
+        - name: PoolMetadataRegistry
+          file: ./abis/PoolMetadataRegistry.json
+      eventHandlers:
+        - event: PoolMetadataUpdated(indexed bytes32,string)
+          handler: handlePoolMetadataUpdated
+      file: ./src/mappings/pool-metadata-registry.ts
```
Comment (ribeirojose) [inline packages/subgraph/manifest.template.yaml:26]: just want to make sure kebab case is the conventional here @mendesfabio ?

File: packages/subgraph/schema.graphql (hunk 2 above)
Comment (ribeirojose) [reply packages/subgraph/schema.graphql:12]: `#question` @mendesfabio can you query subgraphs' gql API using params like `first: number, last: number` etc?

File: packages/subgraph/scripts/generate-manifests.ts [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    await fs.readFile(networksFilePath, { encoding: 'utf-8' })
+  );
+
+  const template = fs.readFileSync('manifest.template.yaml').toString();
+  Object.entries(networks).forEach(([network, config]) => {
+    fs.writeFileSync(
+      `subgraph${network === 'mainnet' ? '' : `.${network}`}.yaml`,
+      Handlebars.compile(template)(config)
+    );
+  });
```
Comment (ribeirojose) [inline packages/subgraph/scripts/generate-manifests.ts:19]: 👀 good first step!

File: packages/subgraph/schema.graphql (hunk 2 above)
Comment (mendesfabio) [reply packages/subgraph/schema.graphql:12]: > `#question` @mendesfabio can you query subgraphs' gql API using params like `first: number, last: number` etc?

@ribeirojose yea they provide a few filters: https://thegraph.com/docs/en/querying/graphql-api/#queries

File: packages/subgraph/manifest.template.yaml (hunk 15 above)
Comment (mendesfabio) [reply packages/subgraph/manifest.template.yaml:26]: not sure if there's a convention but I'm used to camel case.

File: packages/subgraph/.gitignore [hunk 17]
```diff
@@ -0,0 +1,19 @@
+# Geenrated types
+generated/
```
Comment (mendesfabio) [inline packages/subgraph/.gitignore:2]: it seems you changed the `codegen` to output the generated files to `src/types` so this isn't needed anumore.

File: packages/subgraph/scripts/generate-manifests.ts (hunk 13 above)
Comment (mendesfabio) [reply packages/subgraph/scripts/generate-manifests.ts:3]: why did you need it? 😅

Comment (ribeirojose) [reply packages/subgraph/scripts/generate-manifests.ts:3]: dunno 🤫 think I was trying to show @luizakp something

File: packages/subgraph/tests/pool-metadata-registry.test.ts (hunk 10 above)
Comment (mtullyoc) [reply packages/subgraph/tests/pool-metadata-registry.test.ts:12]: Sure... I'll be fixing it on the next [issue](https://linear.app/bleu-llc/issue/BAL-36/create-mapping-for-handling-metadata-updates-events)

File: packages/subgraph/tsconfig.json (hunk 12 above)
Comment (mtullyoc) [reply packages/subgraph/tsconfig.json:11]: It's the auto-generated types folder... Should we preserve it on the compilation?

Comment (mtullyoc) [reply packages/subgraph/tsconfig.json:11]: I'll restore the inheritance:

```
{
  "compilerOptions": {
    "lib": ["DOM","ES2018"],
    "types": ["@graphprotocol/graph-ts", "node"],
    "typeRoots" : ["./node_modules/@types"],
    "skipLibCheck": true,                     /* Skip type checking of declaration files. */
    "forceConsistentCasingInFileNames": true,  /* Disallow inconsistently-cased references to the same file. */
    "strictNullChecks": true
  },
  "exclude": [
    "src/types",
    "node_modules"
  ],
  "extends": "@balancer-pool-metadata/tsconfig/base.json"
}

```

File: packages/subgraph/manifest.template.yaml (hunk 15 above)
Comment (mtullyoc) [reply packages/subgraph/manifest.template.yaml:26]: I changed to camelCase ✌🏻

File: packages/contracts/patch_balancer_contracts [hunk 18]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
```
Comment (ribeirojose) [inline packages/contracts/patch_balancer_contracts:1]: needs rebase

File: packages/subgraph/package.json (hunk 8 above)
Comment (mtullyoc) [reply packages/subgraph/package.json]: Yep! 🎉

Captura de Tela 2023-03-03 às 16 09 55

Comment (mtullyoc) [reply packages/subgraph/package.json]: Sure... on linear [BAL-91](https://linear.app/bleu-llc/issue/BAL-91/create-a-readme-with-anvil-graph-node-setup-instructions)

**Review by mendesfabio — APPROVED** [review]

**Pull request merged**